compiler/src/dmd/common/smallbuffer.d

```diff
@@ -107,28 +107,30 @@ unittest
 }
 
 /**
-(Windows only) Converts a narrow string to a wide string using `buffer` as strorage. Returns a slice managed by
-`buffer` containing the converted string. The terminating zero is not part of the returned slice,
-but is guaranteed to follow it.
+ * (Windows only) Converts a narrow string to a wide string using `buffer` as strorage.
+ * Params:
+ *    narrow = string to be converted
+ *    buffer = where to place the converted string
+ * Returns: a slice of `buffer` containing the converted string. A zero follows the slice.
 */
 version(Windows) wchar[] toWStringz(scope const(char)[] narrow, ref SmallBuffer!wchar buffer) nothrow
 {
-    import core.sys.windows.winnls : MultiByteToWideChar;
-    import dmd.common.file : CodePage;
-
     if (narrow is null)
         return null;
 
-    size_t length;
-    int i;
-    while (1)
+    size_t charsToWchars(scope const(char)[] narrow, scope wchar[] buffer)
     {
         // https://learn.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-multibytetowidechar
-        length = MultiByteToWideChar(CodePage, 0, narrow.ptr, cast(int) narrow.length, buffer.ptr, cast(int) buffer.length);
-        if (length < buffer.length)
-            break;
-        buffer.create(length + 1);
-        assert(++i == 1);   // ensure loop should only execute once or twice
+        import core.sys.windows.winnls : MultiByteToWideChar, CP_ACP;
+        return MultiByteToWideChar(CP_ACP, 0, narrow.ptr, cast(int) narrow.length, buffer.ptr, cast(int) buffer.length);
```

Comment:
This switch to a hardcoded `CP_ACP` here (instead of `dmd.common.file.CodePage`) is problematic for LDC, which uses `CP_UTF8`.

DMD on Windows treats narrow strings as encoded in the current code page, not as proper UTF-8 strings; here for a generic string helper (bad!), and for I/O via the wide Windows API. IIRC, it has to, until all usages of narrow cmdline args and environment variables are banished, switching to a `wmain` entry point and using the wide API for env variables etc. [I've switched LDC to fully UTF-8 only a while back.]